activerecord/lib/active_record/relation/batches.rb

```diff
@@ -287,15 +287,38 @@ def apply_limits(relation, start, finish, order)
       end
 
       def apply_start_limit(relation, start, order)
-        relation.where(predicate_builder[primary_key, start, order == :desc ? :lteq : :gteq])
+        batch_condition(relation, primary_key, start, order == :desc ? :lteq : :gteq)
       end
 
       def apply_finish_limit(relation, finish, order)
-        relation.where(predicate_builder[primary_key, finish, order == :desc ? :gteq : :lteq])
+        batch_condition(relation, primary_key, finish, order == :desc ? :gteq : :lteq)
+      end
+
+      def batch_condition(relation, columns, values, operator)
+        columns = Array(columns)
+        values = Array(values)
+        last_index = columns.size - 1
+        where_clause = nil
+
+        last_index.downto(0) do |column_index|
+          column = columns[column_index]
+          value = values.fetch(column_index)
+
+          where_clause =
+            if column_index == last_index
```

Comment:
Personally I find solutions based on indexes harder to read and maintain. Most likely it performs better on micro benchmarks though a bit more verbose solution should pay off in the long term. For example for a moment I mistakenly read `column_index == last_index` as "the last thing to happen" while it's actually the first thing to happen as we are going from the last index down to 0. Not a big deal when you look closer but still hurts readability. 

So generally I prefer solutions that shape data in a way where we can name things by using terms from the context and not general terms like an index. 

Would you be open to a solution that avoid using indexes and do something like this instead:
```ruby
        cursor_positions = columns.zip(values).reverse!

        first_clause_column, first_clause_value = cursor_positions.pop
        where_clause = predicate_builder[first_clause_column, first_clause_value, operator]

        where_clause = cursor_positions.each do |column_name, value|
          predicate_builder[column_name, value, operator == :lteq ? :lt : :gt].or(
            predicate_builder[column_name, value, :eq].and(where_clause)
          )
        end
```

Definitely open for different naming suggestions but this approach opens more opportunities to give every part of the algorithm a meaningful name. Though to be fair I couldn't come up with a name for the `reverse!` call so end up just inlining it. There is also an option to write it using `inject` but I found it less readable.

Reply:
Both solutions looks fine to me, but yeah, no problems, changed to the suggested. 👍